File: Counting Sort Algorithm/Go/Counting-Sort-Algortihm.go
package main

import (
	"fmt"
)

func counting_sort(arr []int) {
	max, min := arr[0], arr[0]

	for _, num := range arr {
		if num > max {
			max = num
		}
		if num < min {
			min = num
		}
	}

	rangeLen := max - min + 1
	count := make([]int, rangeLen)
	output := make([]int, len(arr))

	for _, num := range arr {
		count[num-min]++
	}

	for i := 1; i < rangeLen; i++ {
		count[i] += count[i-1]
	}

	for i := len(arr) - 1; i >= 0; i-- {
		output[count[arr[i]-min]-1] = arr[i]
		count[arr[i]-min]--
	}

	copy(arr, output)
}

func main() {
	arr := []int{4, 2, 2, 8, 3, 3, 1}

	counting_sort(arr)

	for _, num := range arr {
		fmt.Printf("%d ", num)
	}
}

File: Counting Sort Algorithm/Python/Counting-Sort-Algortihm.py
def counting_sort(arr):
    max_val, min_val = max(arr), min(arr)

    range_len = max_val - min_val + 1
    count = [0] * range_len
    output = [0] * len(arr)

    for num in arr:
        count[num - min_val] += 1

    for i in range(1, range_len):
        count[i] += count[i - 1]

    for i in range(len(arr) - 1, -1, -1):
        output[count[arr[i] - min_val] - 1] = arr[i]
        count[arr[i] - min_val] -= 1

    for i in range(len(arr)):
        arr[i] = output[i]

arr = [4, 2, 2, 8, 3, 3, 1]

counting_sort(arr)

print(arr)
